### PR TITLE
rgw: decrypt multipart get part when encrypted

### PR DESCRIPTION
**Pull request by sungjoon-koh**
If the multipart object is encrypted by SSE-C or SSE-S3, get object with partNumber doesn't decrypt the part.

This patch enables rgw to decrypt the part for get object request with partNumber.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by sungjoon-koh** [comment]
@cbodley Hello, could you please check this PR?

**Comment by sungjoon-koh** [comment]
To reproduce:

1. Enable SSE-S3 (add following line to vstart.sh)
`rgw crypt default encryption key = 4YSmvJtBv0aZ7geVgAsdpRnLBEwWSWlMIGnRS8a9TSA=
`
2. Create multipart, upload part, and complete multipart upload.
3. Get object with part number.
`aws s3api get-object --bucket $BUCKET --key $KEY  --part-number 1 test-part-1`

--> You can get encrypted part.

**Review by cbodley — COMMENTED** [review]
thanks @sungjoon-koh! i've created https://tracker.ceph.com/issues/68292 to track the backport to squid

**Comment by adamemerson** [comment]
jenkins test windows

**Comment by cbodley** [comment]
i added a sse-c test case for this in https://github.com/ceph/s3-tests/pull/591

**Comment by sungjoon-koh** [comment]
@cbodley Thank you for your prompt response. I addressed your comment. 
I also checked with your test.

**Comment by sungjoon-koh** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_rados.cc [hunk 1]
```diff
@@ -6809,6 +6809,13 @@ int RGWRados::Object::Read::prepare(optional_yield y, const DoutPrefixProvider *
   RGWBucketInfo& bucket_info = source->get_bucket_info();
 
   if (params.part_num) {
+    map<string, bufferlist> src_attrset;
+    for (auto& iter : astate->attrset) {
+        if (boost::algorithm::starts_with(iter.first, RGW_ATTR_CRYPT_PREFIX)) {
+            ldpp_dout(dpp, 4) << "get src crypt attr: " << iter.first << dendl;
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_rados.cc:6815]: please stick to 2-space indent to be consistent with https://github.com/ceph/ceph/blob/main/CodingStyle#L76-L77

**Comment by sungjoon-koh** [comment]
jenkins test windows

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-10-04_16:44:30-rgw-wip-68292-distro-default-smithi/ against test cases from https://github.com/ceph/s3-tests/pull/591:
```
s3tests_boto3/functional/test_s3.py::test_multipart_get_part PASSED      [ 67%]
s3tests_boto3/functional/test_s3.py::test_multipart_sse_c_get_part PASSED [ 67%]
s3tests_boto3/functional/test_s3.py::test_multipart_single_get_part PASSED [ 68%]
s3tests_boto3/functional/test_s3.py::test_non_multipart_get_part PASSED  [ 68%]
s3tests_boto3/functional/test_s3.py::test_non_multipart_sse_c_get_part PASSED [ 68%]
```

**Pull request merged**